src/hooks/UseForm/useForm/utils/internal.ts

```diff
@@ -0,0 +1,176 @@
+import React from 'react'
+import { CoordinatesOrNever, Tuple } from '../../../UseGraph/Graph'
+import {
+  FieldData,
+  FieldElement,
+  Fields,
+  FieldsData,
+  FormData,
+  ObjectKey,
+  PartialDataKeys,
+  RegisterResult,
+  isCheckboxInput,
+  isDateInput,
+  isFileInput,
+  isNumberInput,
+  isRadioInput,
+} from '../types'
+
+export const getElementDefaultValue = <TData extends FieldsData>(
+  // the type does not matter since we check for each property
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  element: any,
+): TData[keyof TData] | undefined => {
+  if (!element) return undefined
+
+  if (isRadioInput(element) || isCheckboxInput(element)) {
+    // @ts-expect-error if it is a radio or checkbox the TData[keyof TData] must be a boolean
+    return element.defaultChecked ?? element.checked
+  }
+  if (isFileInput(element)) {
+    // @ts-expect-error if it is a file input the TData[keyof TData] must be files
+    return element.files
+  }
+  if (isNumberInput(element)) {
+    // @ts-expect-error if it is a number input TData[keyof TData] must be a number
+    return element.defaultValue ? +element.defaultValue : +element.value
+  }
+
+  return element.defaultValue ?? element.value
+}
+
+export const getOnChangeValue = <TData extends FieldsData>(
+  // the type does not matter since we check for each property
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  event: any,
+): TData[keyof TData] | undefined => {
+  if ('target' in event && 'type' in event.target) {
+    const target = event.target
+
+    if (isRadioInput(target) || isCheckboxInput(target)) {
+      // @ts-expect-error if it is a radio or checkbox TData[keyof TData] must be a boolean
+      return target.checked
+    }
+    if (isFileInput(target)) {
+      // @ts-expect-error if it is a file input TData[keyof TData] must be files
+      return target.files
+    }
+    if (isNumberInput(target)) {
+      // @ts-expect-error if it is a number input TData[keyof TData] must be a number
+      return +target.value
+    }
+    return target.value
+  }
+}
+
+export const isBlank = (value: unknown): value is undefined | null | '' =>
+  value === null || value === undefined || value === ''
+
+export const isNotBlank = (value: unknown): value is Exclude<unknown, undefined | null | ''> =>
+  !isBlank(value)
```

Comment:
Not sure that TypeScript actually does a good job of excluding from a large primitive type like `unknown`. Instead, we can add a generic to make sure that the type guard does the narrowing that we expect.


```suggestion
export const isNotBlank = <T>(value: T): value is Exclude<T, undefined | null | ''> =>
  !isBlank(value)
```

Reply:
FYI, the generic `T` may need to extend from some field value type if you want `isNotBlank` to be coupled with form field logic